File: server/app/routes/events.py
"""이벤트 조회와 수동 정정 (API명세서 §4.1 · FN-REC-04 · FN-EVT-05).

* `GET /events` · `GET /events/{event_id}` — 목록과 상세
* `PATCH /events/{event_id}` — 오탐 표시 · 강제 종결

클립(`clip_url`)과 키프레임은 아직 비어 있다 — 예약 추출이 M4(FN-REC-03)다. 그 자리를
그럴듯한 값으로 메우지 않는다.
"""

from __future__ import annotations

import logging
from typing import Annotated, Protocol

from fastapi import APIRouter, HTTPException, Query, Request
from pydantic import AwareDatetime

from aegis_contracts import (
    ErrorBody,
    ErrorResponse,
    EventDetail,
    EventListQuery,
    EventListResponse,
    EventPatchRequest,
    EventStatus,
    EventSummary,
    ViolationType,
)
from server.app.event_service import EventService

__all__ = ["router"]

log = logging.getLogger("server.routes.events")

router = APIRouter(tags=["events"])


class EventReader(Protocol):
    """이 라우터가 요구하는 조회 능력. 구현은 `server/infra/db/repository.py`."""

    async def list_events(self, query: EventListQuery) -> tuple[list[EventSummary], str | None]: ...

    async def get(self, event_id: str) -> EventDetail | None: ...


@router.get(
    "/events",
    response_model=EventListResponse,
    responses={400: {"model": ErrorResponse}, 503: {"model": ErrorResponse}},
)
async def list_events(
    request: Request,
    from_: Annotated[AwareDatetime | None, Query(alias="from")] = None,
    to: AwareDatetime | None = None,
    cam_id: int | None = None,
    type: ViolationType | None = None,  # noqa: A002 — §4.1 이 정한 쿼리 파라미터 이름이다
    status: EventStatus | None = None,
    zone_id: str | None = None,
    limit: int | None = None,
    cursor: str | None = None,
) -> EventListResponse:
    """이벤트 목록. 쿼리 파라미터는 §4.1 그대로다."""
    query = EventListQuery.model_validate(
        {
            "from": from_,
            "to": to,
            "cam_id": cam_id,
            "type": type,
            "status": status,
            "zone_id": zone_id,
            "limit": limit,
            "cursor": cursor,
        }
    )
    reader = _reader(request)
    try:
        items, next_cursor = await reader.list_events(query)
    except ValueError as exc:
        # 깨진 커서. 조용히 첫 장으로 되돌리면 클라이언트가 같은 장을 무한히 돈다.
        raise _error(400, "VALIDATION_ERROR", str(exc), {"cursor": cursor}) from exc
    except OSError as exc:
        raise _unavailable(exc) from exc
    return EventListResponse(items=items, next_cursor=next_cursor)


@router.get(
    "/events/{event_id}",
    response_model=EventDetail,
    responses={404: {"model": ErrorResponse}, 503: {"model": ErrorResponse}},
)
async def get_event(request: Request, event_id: str) -> EventDetail:
    try:
        event = await _reader(request).get(event_id)
    except OSError as exc:
        raise _unavailable(exc) from exc
    if event is None:
        raise _error(404, "NOT_FOUND", "요청한 이벤트가 존재하지 않습니다", {"event_id": event_id})
    return event


@router.patch(
    "/events/{event_id}",
    response_model=EventDetail,
    responses={404: {"model": ErrorResponse}, 503: {"model": ErrorResponse}},
)
async def patch_event(
    request: Request,
    event_id: str,
    body: EventPatchRequest,
) -> EventDetail:
    """FN-EVT-05 — 수동 정정. 기능명세서 §4.2 · API명세서 §4.1

    | 요청 | 뜻 |
    |---|---|
    | `is_false_positive` | 오탐이었다. **지표에서 전량 제외**된다(§6.7) |
    | `force_resolve` | 시스템이 놓친 시정을 관리자가 종결한다 |

    **`fall` 은 관리자 확인으로 종결하는 것이 기본 절차다** — 쓰러진 사람은 방송을
    듣고 스스로 시정할 수 없으므로 자동 해소 조건이 성립하지 않는다.
    """
    service: EventService | None = getattr(request.app.state, "event_service", None)
    if service is None:
        raise _error(503, "NOT_FOUND", "이벤트 저장소가 연결되지 않았습니다")
    try:
        event = await service.patch(event_id, body)
    except OSError as exc:
        raise _unavailable(exc) from exc
    if event is None:
        raise _error(404, "NOT_FOUND", "요청한 이벤트가 존재하지 않습니다", {"event_id": event_id})
    return event


def _reader(request: Request) -> EventReader:
    reader: EventReader | None = getattr(request.app.state, "events", None)
    if reader is None:
        # 조립이 잘못된 것이지 사용자의 잘못이 아니다. 빈 목록으로 덮지 않는다 —
        # "이벤트가 없다"와 "저장소가 없다"는 화면에서 구분되어야 한다.
        raise _error(503, "NOT_FOUND", "이벤트 저장소가 연결되지 않았습니다")
    return reader


def _unavailable(exc: Exception) -> HTTPException:
    log.warning("이벤트 저장소에 닿지 못했다: %s", exc)
    return _error(503, "NOT_FOUND", "이벤트 저장소에 닿지 못했습니다", {"reason": str(exc)})


def _error(
    http_status: int,
    code: str,
    message: str,
    detail: dict[str, object] | None = None,
) -> HTTPException:
    """§1.4 오류 봉투. FastAPI 기본 `{"detail": ...}` 형태를 쓰지 않는다."""
    body = ErrorResponse(
        error=ErrorBody.model_validate({"code": code, "message": message, "detail": detail})
    )
    return HTTPException(status_code=http_status, detail=body.model_dump(mode="json"))

File: front/src/api/events.ts
/**
 * 이벤트 조회와 수동 정정 (API명세서 §4.1 · FN-UI-03 · FN-EVT-05).
 *
 * 목록은 **커서 페이징**이다(`next_cursor`). 오프셋 페이징을 쓰면 새 이벤트가 들어오는
 * 동안 페이지 경계가 밀려 같은 이벤트가 두 번 보이거나 빠진다.
 */

import type { EventDetail, EventListResponse, EventStatus, ViolationType } from '../types/system'
import { getJson, queryString, sendJson } from './client'

/** `GET /events` 쿼리(§4.1). 화면이 쓰는 것만 노출한다. */
export type EventQuery = {
  from?: string | null
  to?: string | null
  cam_id?: number | null
  type?: ViolationType | null
  status?: EventStatus | null
  zone_id?: string | null
  limit?: number | null
  cursor?: string | null
}

export async function fetchEvents(
  query: EventQuery = {},
  signal?: AbortSignal,
): Promise<EventListResponse> {
  return getJson<EventListResponse>(`/events${queryString({ ...query })}`, signal)
}

export async function fetchEvent(eventId: string, signal?: AbortSignal): Promise<EventDetail> {
  return getJson<EventDetail>(`/events/${encodeURIComponent(eventId)}`, signal)
}

/**
 * `PATCH /events/{id}` — 오탐 표시 · 강제 종결 · 메모(§4.1 · FN-EVT-05).
 *
 * `Partial` 을 쓴다. 계약 모델은 세 필드를 전부 요구하지만(생성기가 직렬화 기준으로
 * 낸다) **일부만 보내는 것이 이 요청의 용법**이다 — 서버는 `None` 인 필드를 건드리지
 * 않는다. 메모만 고치려는데 `is_false_positive: null` 을 함께 보내야 하는 것은
 * 계약이 아니라 생성기의 부작용이다.
 */
export type EventPatch = {
  is_false_positive?: boolean
  note?: string
  force_resolve?: boolean
}

export async function patchEvent(
  eventId: string,
  patch: EventPatch,
  signal?: AbortSignal,
): Promise<EventDetail | null> {
  return sendJson<EventDetail>('PATCH', `/events/${encodeURIComponent(eventId)}`, patch, signal)
}

/**
 * `GET /events/{id}/clip` 의 URL. 화면은 `<video src>` 로 그대로 쓴다.
 *
 * `clip_url`(`/media/clips/...`)이 아니라 이 경로를 쓰는 이유: `clip_status` 가
 * `pending` 인 동안 서버가 **404 로 없는 것을 없다고** 말해준다(§4.1). 정적 경로로
 * 직접 붙으면 그 구분이 없어 브라우저가 조용히 재생에 실패한다.
 */
export function clipSrc(eventId: string): string {
  return `/api/v1/events/${encodeURIComponent(eventId)}/clip`
}
